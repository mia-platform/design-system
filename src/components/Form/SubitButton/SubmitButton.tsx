import { ReactElement, ReactNode, useCallback } from 'react'
import { FormInstance } from 'antd'

import { Button } from '../../Button'
import { ButtonProps } from '../../Button/Button.props.ts'
import { FormItem } from '../FormItem/FormItem.tsx'

export type SubmitButtonProps = Omit<ButtonProps, 'children'> & {
  children?: ReactNode | ((form: FormInstance) => ReactNode)
}

const defaultSubmitButton = (
  label: ReactNode = 'Submit',
  props: ButtonProps = {}
): ReactNode => {
  return (
    <Button htmlType={Button.HTMLType.Submit} {...props}>
      {label}
    </Button>
  )
}

export const SubmitButton = (
  {
    children,
    ...props
  }: SubmitButtonProps
): ReactElement => {
  const renderSubmitButton = useCallback(({ form }: { form: FormInstance }) => {
    if (typeof children === 'function') {
      return children(form)
    }
    if (typeof children === 'boolean') {
      return children && defaultSubmitButton()
    }
    return defaultSubmitButton(children, props)
  }, [children, props])

  return (
    <FormItem isFullWidth justify="end">
      {renderSubmitButton}
    </FormItem>
  )
}
