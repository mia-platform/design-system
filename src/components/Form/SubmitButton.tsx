import { ReactElement, ReactNode, useCallback } from 'react'
import { FormInstance } from 'antd'

import { Button } from '../Button'
import { FormItem } from './FormItem.tsx'

export type SubmitButtonProps = {
  children?: ReactNode | ((form: FormInstance) => ReactNode)
}

const defaultSubmitButton = (label: ReactNode = 'Submit'): ReactNode => {
  return (
    <Button htmlType={Button.HTMLType.Submit}>
      {label}
    </Button>
  )
}

export const SubmitButton = (
  {
    children,
  }: SubmitButtonProps
): ReactElement => {
  const renderSubmitButton = useCallback(({ form }: { form: FormInstance }) => {
    if (typeof children === 'function') {
      return children(form)
    }
    if (typeof children === 'boolean') {
      return children && defaultSubmitButton()
    }
    return defaultSubmitButton(children)
  }, [children])

  return (
    <FormItem isFullWidth justify="end">
      {renderSubmitButton}
    </FormItem>
  )
}
