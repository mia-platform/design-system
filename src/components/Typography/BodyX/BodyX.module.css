/**
 * Copyright 2023 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

.bodyS {
  margin: 0 !important;
  font-size: var(--typography-bodyS-fontSize, 14px);
  line-height: var(--typography-bodyS-lineHeight, 16px);
}

.bodyM {
  margin: 0 !important;
  font-size: var(--typography-bodyM-fontSize, 16px);
  line-height: var(--typography-bodyM-lineHeight, 24px);
}

.bodyL {
  margin: 0 !important;
  font-size: var(--typography-bodyL-fontSize, 18px);
  line-height: var(--typography-bodyL-lineHeight, 24px);
}

.bold {
  color: var(--palette-text-neutral-bold, #464646);
}

.subtle {
  color: var(--palette-text-neutral-subtle, #898989);
}
