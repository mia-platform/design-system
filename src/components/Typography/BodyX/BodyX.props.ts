/**
 * Copyright 2023 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { ReactNode } from 'react'

import { CopyConfig, EllipsisConfig } from '../Typography.types'

export type BodyXProps = {

  /**
   * The children nodes to be rendered within the typography context.
   */
  children?: ReactNode,

  /**
   * Whether the text is copyable. If set to true, a copy icon is shown to the right of the text.
   * The attribute is further customizable according to the following data model:
   *
   * `object`:
   *   - text?: The text to copy. <br>`string`
   *   - onCopy?: Function called when the text is copied. <br>`() => void`
   *   - icon?: Custom copy icon(s).
   *   It is possible to set a single icon that will become the new copyIcon,
   *   or an array of 2 icons that will replace both copyIcon and copiedIcon. <br>`ReactNode`
   *   - tooltips?: Custom tooltip text(s).
   *   It is possible to set a single string that will become the new copyText,
   *   or an array of 2 strings that will replace both copyText and copiedText.
   *   If the attribute is set to false, no tooltip will be shown. <br>`boolean | ReactNode`
   *   - format?: The Mime Type of the text. <br>`'text/plain' | 'text/html'`
   */
  copyable?: boolean | CopyConfig,

  /**
   * Displays ellipsis when text overflows.
   * The attribute is customizable according to the following data model:
   *
   * `object`:
   *   - rows?: The maximum number of rows the content can occupy before
   *   going into ellipsis or being truncated. <br>`number`
   *   - expandable?: Whether the text is expandable. <br>`boolean`
   *   - suffix?: Suffix of the ellipsis content. <br>`string`
   *   - symbol?: Custom description of the ellipsis (default is 'Expand'). <br>`ReactNode`
   *   - onExpand?: Function called when the content is expanded. <br>`MouseEventHandler<HTMLElement>`
   *   - onEllipsis?: Function called when the content enters or leaves the ellipsis state. <br>`() => void`
   *   - tooltip?: Whether to display a tooltip on ellipsis.
   *   If set to true, a copy icon is shown to the right of the text.
   *   The attribute is further customizable. <br>`ReactNode`
   */
  ellipsis?: boolean | EllipsisConfig,

  /**
   * Whether the body font weight is bold.
   */
  isBold?: boolean,
}
