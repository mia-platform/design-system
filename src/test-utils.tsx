/**
 * Copyright 2023 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { RenderOptions, RenderResult, render } from '@testing-library/react'
import { ConfigProvider } from 'antd'
import { ReactNode } from 'react'

import { ThemeContext, defaultTheme } from './components/ThemeProvider'
import { ANT_PREFIX } from './components/ThemeProvider/Ant'

/**
 * Wraps a React node with a series of providers for testing purposes.
 *
 * @remarks It adds the default theme to components
 * @remarks It adds the default Ant class prefix
 * @remarks It removes the css-do-not-override classes from Ant Design components
 *
 * @param {ReactNode} children - The React node to be wrapped.
 * @returns {ReactNode} The wrapped React node.
 */
const TestWrapper = ({ children }: {children: ReactNode}): ReactNode => (
  <ThemeContext.Provider value={defaultTheme}>
    <ConfigProvider prefixCls={ANT_PREFIX} theme={{ hashed: false }}>
      {children}
    </ConfigProvider>
  </ThemeContext.Provider>
)

/**
 * Custom render function for testing React components.
 *
 * @param {ReactNode} element - The React node to be rendered.
 * @param {RenderOptions} [options] - Options for rendering.
 * @returns {RenderResult} The result of rendering the React node.
 */
const customRender = (
  element: ReactNode,
  options?: Omit<RenderOptions, 'wrapper'>,
): RenderResult => render(element, { wrapper: TestWrapper, ...options })

export * from '@testing-library/react'
export {
  customRender as render,
  TestWrapper,
}
