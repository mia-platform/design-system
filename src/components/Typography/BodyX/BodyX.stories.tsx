/**
 * Copyright 2023 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import type { Meta, StoryObj } from '@storybook/react'

import { customCopyable, customEllipsis, fontUrl, loremIpsum } from '../Typography.mocks'
import { BodyX } from './BodyX'
import { Size } from './BodyX.types'
import { Typography } from '..'

const { S } = Size

const bodySLongText = ['BodyS', loremIpsum].join(' | ')
const bodyMLongText = ['BodyM', loremIpsum].join(' | ')
const bodyLLongText = ['BodyL', loremIpsum].join(' | ')
const bodySBoldLongText = ['BodyS Bold', loremIpsum].join(' | ')
const bodyMBoldLongText = ['BodyM Bold', loremIpsum].join(' | ')
const bodyLBoldLongText = ['BodyL Bold', loremIpsum].join(' | ')

const bodySEllipsisWithTooltip = { rows: 3, tooltip: bodySLongText }
const bodyMEllipsisWithTooltip = { rows: 3, tooltip: bodyMLongText }
const bodyLEllipsisWithTooltip = { rows: 3, tooltip: bodyLLongText }
const bodySBoldEllipsisWithTooltip = { rows: 3, tooltip: bodySBoldLongText }
const bodyMBoldEllipsisWithTooltip = { rows: 3, tooltip: bodyMBoldLongText }
const bodyLBoldEllipsisWithTooltip = { rows: 3, tooltip: bodyLBoldLongText }

const meta = {
  title: 'Components/Typography/BodyX',
  component: BodyX,
  parameters: {
    design: { type: 'iframe', name: 'Font Definition', url: fontUrl },
  },
  args: {
    ...BodyX.defaultProps,
    children: 'Text',
  },
  argTypes: {
    copyable: { control: 'boolean' },
    ellipsis: { control: 'boolean' },
    isBold: { table: { disable: true } },
    size: { table: { disable: true } },
  },
} satisfies Meta<typeof BodyX>

export default meta
type Story = StoryObj<typeof meta>

export const BodyXComponent: Story = {
  args: { ...meta.args, size: S },
  argTypes: {
    isBold: { table: { disable: false } },
    size: { table: { disable: false } },
  },
}

export const All = {
  argTypes: {
    children: { table: { disable: true } },
  },
  render: (args: object) => (
    <>
      <Typography.BodyS {...args}>{'BodyS'}</Typography.BodyS>
      <Typography.BodyS {...args} isBold>{'BodyS Bold'}</Typography.BodyS>
      <Typography.BodyM {...args}>{'BodyM'}</Typography.BodyM>
      <Typography.BodyM {...args} isBold>{'BodyM Bold'}</Typography.BodyM>
      <Typography.BodyL {...args}>{'BodyL'}</Typography.BodyL>
      <Typography.BodyL {...args} isBold>{'BodyL Bold'}</Typography.BodyL>
    </>
  ),
}

export const BodyS = {
  render: (args: object) => <Typography.BodyS {...args} />,
}

export const BodySBold = {
  render: (args: object) => <Typography.BodyS {...args} isBold />,
}

export const BodyM = {
  render: (args: object) => <Typography.BodyM {...args} />,
}

export const BodyMBold = {
  render: (args: object) => <Typography.BodyM {...args} isBold />,
}

export const BodyL = {
  render: (args: object) => <Typography.BodyL {...args} />,
}

export const BodyLBold = {
  render: (args: object) => <Typography.BodyL {...args} isBold />,
}

export const WithEllipsisWithTooltip = {
  argTypes: {
    children: { table: { disable: true } },
  },
  render: (args: object) => (
    <>
      <Typography.BodyS {...args} ellipsis={bodySEllipsisWithTooltip}>{bodySLongText}</Typography.BodyS>
      <Typography.BodyS {...args} ellipsis={bodySBoldEllipsisWithTooltip} isBold>{bodySBoldLongText}</Typography.BodyS>
      <Typography.BodyM {...args} ellipsis={bodyMEllipsisWithTooltip}>{bodyMLongText}</Typography.BodyM>
      <Typography.BodyM {...args} ellipsis={bodyMBoldEllipsisWithTooltip} isBold>{bodyMBoldLongText}</Typography.BodyM>
      <Typography.BodyL {...args} ellipsis={bodyLEllipsisWithTooltip}>{bodyLLongText}</Typography.BodyL>
      <Typography.BodyL {...args} ellipsis={bodyLBoldEllipsisWithTooltip} isBold>{bodyLBoldLongText}</Typography.BodyL>
    </>
  ),
}

export const WithEllipsisCustom = {
  args: { ...meta.args, ellipsis: customEllipsis },
  argTypes: {
    children: { table: { disable: true } },
  },
  render: (args: object) => (
    <>
      <Typography.BodyS {...args}>{bodySLongText}</Typography.BodyS>
      <Typography.BodyS {...args} isBold>{bodySBoldLongText}</Typography.BodyS>
      <Typography.BodyM {...args}>{bodyMLongText}</Typography.BodyM>
      <Typography.BodyM {...args} isBold>{bodyMBoldLongText}</Typography.BodyM>
      <Typography.BodyL {...args}>{bodyLLongText}</Typography.BodyL>
      <Typography.BodyL {...args} isBold>{bodyLBoldLongText}</Typography.BodyL>
    </>
  ),
}

export const Copyable = {
  args: { ...meta.args, copyable: true },
  argTypes: {
    children: { table: { disable: true } },
  },
  render: (args: object) => (
    <>
      <Typography.BodyS {...args}>{'BodyS'}</Typography.BodyS>
      <Typography.BodyS {...args} isBold>{'BodyS Bold'}</Typography.BodyS>
      <Typography.BodyM {...args}>{'BodyM'}</Typography.BodyM>
      <Typography.BodyM {...args} isBold>{'BodyM Bold'}</Typography.BodyM>
      <Typography.BodyL {...args}>{'BodyL'}</Typography.BodyL>
      <Typography.BodyL {...args} isBold>{'BodyL Bold'}</Typography.BodyL>
    </>
  ),
}

export const CopyableCustom = {
  args: { ...meta.args, copyable: customCopyable },
  argTypes: {
    children: { table: { disable: true } },
  },
  render: (args: object) => (
    <>
      <Typography.BodyS {...args}>{'BodyS'}</Typography.BodyS>
      <Typography.BodyS {...args} isBold>{'BodyS Bold'}</Typography.BodyS>
      <Typography.BodyM {...args}>{'BodyM'}</Typography.BodyM>
      <Typography.BodyM {...args} isBold>{'BodyM Bold'}</Typography.BodyM>
      <Typography.BodyL {...args}>{'BodyL'}</Typography.BodyL>
      <Typography.BodyL {...args} isBold>{'BodyL Bold'}</Typography.BodyL>
    </>
  ),
}
