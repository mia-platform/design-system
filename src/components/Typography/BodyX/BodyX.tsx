/**
 * Copyright 2023 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { ReactElement, useMemo } from 'react'
import { Typography as AntTypography } from 'antd'
import classnames from 'classnames'

import { BodyXSize, Size } from './BodyX.types'
import { BodyXProps } from './BodyX.props'
import styles from './BodyX.module.css'

const { Paragraph: AntParagraph } = AntTypography

const { bodyS, bodyM, bodyL } = styles

const { S, M, L } = Size

/**
 * UI component for displaying bodies (BodyS, BodyM, BodyL).
 *
 * @link https://ant.design/components/typography#typographyparagraph
 * @returns {BodyX} BodyX component
 */
export const BodyX = ({
  children,
  copyable,
  ellipsis,
  isBold,
  size,
}: BodyXProps & BodyXSize): ReactElement => {
  const bodyClassName = useMemo(() => classnames([
    size === S && bodyS,
    size === M && bodyM,
    size === L && bodyL,
  ]), [size])

  return (
    <AntParagraph
      className={bodyClassName}
      copyable={copyable}
      ellipsis={ellipsis}
      role={'paragraph'}
      strong={isBold}
    >
      {children}
    </AntParagraph>
  )
}

BodyX.defaultProps = {
  copyable: false,
  ellipsis: false,
  isBold: false,
}
