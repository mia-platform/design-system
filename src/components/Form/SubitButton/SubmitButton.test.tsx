/**
 * Copyright 2023 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { render, screen } from '../../../test-utils.tsx'
import { BodyM } from '../../Typography/BodyX/BodyM'
import { SubmitButton } from './SubmitButton'

const exampleText = 'test'

describe('SubmitButton Component', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  test('renders correctly with default props', () => {
    const { asFragment } = render(<SubmitButton />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('renders correctly with a custom label passed as child', () => {
    const { asFragment } = render(
      <SubmitButton>
        <BodyM>{exampleText}</BodyM>
      </SubmitButton>
    )
    expect(asFragment()).toMatchSnapshot()
    expect(screen.getByText(exampleText)).toBeInTheDocument()
  })

  test('renders correctly with a render function passed as child', () => {
    const { asFragment } = render(
      <SubmitButton>
        {() => <BodyM>{exampleText}</BodyM>}
      </SubmitButton>
    )
    expect(asFragment()).toMatchSnapshot()
    expect(screen.getByText(exampleText)).toBeInTheDocument()
  })
})
