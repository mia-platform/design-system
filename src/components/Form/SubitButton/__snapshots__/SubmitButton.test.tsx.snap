// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubmitButton Component renders correctly with a custom label passed as child 1`] = `
<DocumentFragment>
  <div
    class="mia-platform-form-item"
    style="margin: 0px; justify-self: end; grid-column: -1 / 1;"
  >
    <div
      class="mia-platform-row mia-platform-form-item-row"
    >
      <div
        class="mia-platform-col mia-platform-form-item-control"
      >
        <div
          class="mia-platform-form-item-control-input"
        >
          <div
            class="mia-platform-form-item-control-input-content"
          >
            <button
              class="mia-platform-btn mia-platform-btn-primary mia-platform-btn-color-primary mia-platform-btn-variant-solid button buttonMd"
              type="submit"
            >
              <div
                class="buttonText"
              >
                <div
                  class="mia-platform-typography bodyM main"
                  role="paragraph"
                >
                  test
                </div>
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SubmitButton Component renders correctly with a render function passed as child 1`] = `
<DocumentFragment>
  <div
    class="mia-platform-form-item"
    style="margin: 0px; justify-self: end; grid-column: -1 / 1;"
  >
    <div
      class="mia-platform-row mia-platform-form-item-row"
    >
      <div
        class="mia-platform-col mia-platform-form-item-control"
      >
        <div
          class="mia-platform-form-item-control-input"
        >
          <div
            class="mia-platform-form-item-control-input-content"
          >
            <div
              class="mia-platform-typography bodyM main"
              role="paragraph"
            >
              test
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SubmitButton Component renders correctly with default props 1`] = `
<DocumentFragment>
  <div
    class="mia-platform-form-item"
    style="margin: 0px; justify-self: end; grid-column: -1 / 1;"
  >
    <div
      class="mia-platform-row mia-platform-form-item-row"
    >
      <div
        class="mia-platform-col mia-platform-form-item-control"
      >
        <div
          class="mia-platform-form-item-control-input"
        >
          <div
            class="mia-platform-form-item-control-input-content"
          >
            <button
              class="mia-platform-btn mia-platform-btn-primary mia-platform-btn-color-primary mia-platform-btn-variant-solid button buttonMd"
              type="submit"
            >
              <div
                class="buttonText"
              >
                Submit
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
